greeting.go

```diff
@@ -1,87 +1,89 @@
+	Objects    []string `xml:"svcMenu>objURI"`
+	Extensions []string `xml:"svcMenu>svcExtension>extURI,omitempty"`
 }
 
-// DCPAccess defines the data collection policy for an EPP server.
-type DCPAccess struct {
-	All              Bool `xml:"all"`
-	None             Bool `xml:"none"`
-	Null             Bool `xml:"null"`
-	Personal         Bool `xml:"personal"`
-	PersonalAndOther Bool `xml:"personalAndOther"`
-	Other            Bool `xml:"other"`
+func (c *Conn) readGreeting() error {
+	err := c.readDataUnit()
+	if err != nil {
+		return err
+	}
+	return decodeGreeting(&c.decoder, &c.Greeting)
 }
 
-// DCPStatement defines the data collection policy
-// for an EPP server. The EPP server may return 1 or more
-// DCPStatements in a Greeting message.
-// https://tools.ietf.org/html/rfc5730#section-2.4
-type DCPStatement struct {
-	Purpose struct {
-		Admin        Bool `xml:"admin"`
-		Contact      Bool `xml:"contact"`
-		Provisioning Bool `xml:"prov"`
-		Other        Bool `xml:"other"`
-	} `xml:"purpose"`
-	Recipient struct {
-		Other Bool `xml:"other"`
-		Ours  *struct {
-			Recipient string `xml:"recDesc"`
-		} `xml:"ours"`
-		Public    Bool `xml:"public"`
-		Same      Bool `xml:"same"`
-		Unrelated Bool `xml:"unrelated"`
-	} `xml:"recipient"`
-	Retention struct {
-		Business   Bool `xml:"business"`
-		Indefinite Bool `xml:"indefinite"`
-		Legal      Bool `xml:"legal"`
-		None       Bool `xml:"none"`
-		Stated     Bool `xml:"stated"`
-	} `xml:"retention"`
-	Expiry *struct {
-		Absolute Bool `xml:"absolute"`
-		Relative Bool `xml:"relative"`
-	} `xml:"expiry"`
-}
+func decodeGreeting(d *Decoder, g *Greeting) error {
+	d.Reset()
+	g.ServerName = ""
+	g.Languages = g.Languages[:0]
+	g.Versions = g.Versions[:0]
+	g.Objects = g.Objects[:0]
+	g.Extensions = g.Extensions[:0]
+	for {
+		t, err := d.Token()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		switch node := t.(type) {
+		case xml.StartElement:
+			// Ignore <dcp> section entirely
+			if node.Name.Local == "dcp" {
+				err = d.Skip()
+				if err != nil {
+					return err
+				}
+			}
 
-// A ErrGreetingNotFound is reported when a <greeting> message is expected but not found.
-var ErrGreetingNotFound = errors.New("missing <greeting> message")
+		case xml.EndElement:
+			// Escape early (skip remaining XML)
+			if node.Name.Local == "svcMenu" &&
+				g.ServerName != "" {
+				return nil
+			}
 
-// ReadGreeting reads a <greeting> message from the EPP server.
-// Performed automatically during a Handshake or Hello command.
-func (c *Conn) ReadGreeting() (*Greeting, error) {
-	var msg message
-	err := c.readMessage(&msg)
-	if err != nil {
-		return nil, err
-	}
-	if msg.Greeting == nil {
-		return nil, ErrGreetingNotFound
+		case xml.CharData:
+			e := d.Element()
+			if e == nil {
+				break
+			}
+			switch e.Name.Local {
+			case "svID":
+				g.ServerName = string(node)
+			case "version":
+				g.Versions = append(g.Versions, string(node))
```

Comment:
Do we use these values upstream? Thinking about why we'd need to decode them at all.